### ISSUE TITLE
Incorrect lowering the function type of copy constructor leads to mlir validation failure

### ISSUE THREAD

**Issue by kumarak** [trailofbits/vast#347]
The function type of default copy constructor is getting lowered incorrectly causing the MLIR validation failure due to type conversion issue. Attached generated mlir and source file for reproducing the issue:

error:
```
input.cc:2:9: error: 'hl.ref' op inferred type(s) '!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>' are incompatible with return type(s) of operation '!hl.lvalue<!hl.elaborated<!hl.record<"Student">,  const >>'
```

mlir module dump:
```
"builtin.module"() ({
  "hl.func"() ({
  }) {function_type = (!hl.lvalue<!hl.ptr<!hl.void>>, !hl.lvalue<!hl.ptr<!hl.void< const >>>, !hl.lvalue<!hl.long< unsigned >>) -> !hl.ptr<!hl.void>, linkage = 0 : i32, sym_name = "__builtin_memcpy", sym_visibility = "private"} : () -> ()
  "hl.func"() ({
  ^bb0(%arg0: !hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>):
    "hl.scope"() ({
      %0 = "hl.this"() : () -> !hl.ptr<!hl.record<"Student">>
      %1 = "hl.member"(%0) {name = "registry"} : (!hl.ptr<!hl.record<"Student">>) -> !hl.lvalue<!hl.int>
      %2 = "hl.ref"(%arg0) : (!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>) -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">,  const >>
      %3 = "hl.member"(%2) {name = "registry"} : (!hl.lvalue<!hl.elaborated<!hl.record<"Student">,  const >>) -> !hl.lvalue<!hl.int< const >>
      %4 = "hl.implicit_cast"(%3) {kind = 4 : i64} : (!hl.lvalue<!hl.int< const >>) -> !hl.int
      %5 = "hl.assign"(%4, %1) : (!hl.int, !hl.lvalue<!hl.int>) -> !hl.int
      %6 = "hl.this"() : () -> !hl.ptr<!hl.record<"Student">>
      %7 = "hl.member"(%6) {name = "name"} : (!hl.ptr<!hl.record<"Student">>) -> !hl.lvalue<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>
      %8 = "hl.addressof"(%7) : (!hl.lvalue<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>) -> !hl.ptr<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>
      %9 = "hl.implicit_cast"(%8) {kind = 1 : i64} : (!hl.ptr<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>) -> !hl.ptr<!hl.void>
      %10 = "hl.ref"(%arg0) : (!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>) -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">,  const >>
      %11 = "hl.member"(%10) {name = "name"} : (!hl.lvalue<!hl.elaborated<!hl.record<"Student">,  const >>) -> !hl.lvalue<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>
      %12 = "hl.addressof"(%11) : (!hl.lvalue<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>) -> !hl.ptr<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>
      %13 = "hl.implicit_cast"(%12) {kind = 1 : i64} : (!hl.ptr<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>) -> !hl.ptr<!hl.void< const >>
      %14 = "hl.const"() {value = #hl.integer<40> : !hl.long< unsigned >} : () -> !hl.long< unsigned >
      %15 = "hl.call"(%9, %13, %14) {callee = @__builtin_memcpy} : (!hl.ptr<!hl.void>, !hl.ptr<!hl.void< const >>, !hl.long< unsigned >) -> !hl.ptr<!hl.void>
      %16 = "hl.this"() : () -> !hl.ptr<!hl.record<"Student">>
      %17 = "hl.deref"(%16) : (!hl.ptr<!hl.record<"Student">>) -> !hl.lvalue<!hl.record<"Student">>
      "hl.return"(%17) : (!hl.lvalue<!hl.record<"Student">>) -> ()
    }) : () -> ()
  }) {function_type = (!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>) -> !hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">>>>, linkage = 3 : i32, sym_name = "_ZN7StudentaSERKS_"} : () -> ()
  "hl.cxxstruct"() ({
  ^bb0:
  }, {
    "hl.field"() {name = "registry", type = !hl.int} : () -> ()
    "hl.field"() {name = "name", type = !hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>} : () -> ()
    "hl.func"() ({
      "hl.scope"() ({
        "hl.return"() : () -> ()
      }) : () -> ()
    }) {function_type = () -> (), linkage = 3 : i32, sym_name = "_ZN7StudentC1Ev"} : () -> ()
    "hl.func"() ({
    ^bb0(%arg0: !hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>):
      "hl.return"() : () -> ()
    }) {function_type = (!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>) -> (), linkage = 3 : i32, sym_name = "_ZN7StudentC1ERKS_"} : () -> ()
    "hl.func"() ({
    ^bb0(%arg0: !hl.lvalue<!hl.rvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">>>>>):
      "hl.return"() : () -> ()
    }) {function_type = (!hl.lvalue<!hl.rvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">>>>>) -> (), linkage = 3 : i32, sym_name = "_ZN7StudentC1EOS_"} : () -> ()
  }) {name = "Student"} : () -> ()
  "hl.func"() ({
  ^bb0(%arg0: !hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"Student">>>>, %arg1: !hl.lvalue<!hl.int>):
    "hl.scope"() ({
      %0 = "hl.var"() ({
        %9 = "hl.unsupportedOp"() {name = "CXXConstructExpr"} : () -> !hl.elaborated<!hl.record<"Student">>
        "hl.value.yield"(%9) : (!hl.elaborated<!hl.record<"Student">>) -> ()
      }, {
      }) {name = "aux"} : () -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">>>
      %1 = "hl.ref"(%0) : (!hl.lvalue<!hl.elaborated<!hl.record<"Student">>>) -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">>>
      %2 = "hl.ref"(%arg0) : (!hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"Student">>>>) -> !hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"Student">>>>
      %3 = "hl.implicit_cast"(%2) {kind = 4 : i64} : (!hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"Student">>>>) -> !hl.ptr<!hl.elaborated<!hl.record<"Student">>>
      %4 = "hl.ref"(%arg1) : (!hl.lvalue<!hl.int>) -> !hl.lvalue<!hl.int>
      %5 = "hl.implicit_cast"(%4) {kind = 4 : i64} : (!hl.lvalue<!hl.int>) -> !hl.int
      %6 = "hl.subscript"(%3, %5) : (!hl.ptr<!hl.elaborated<!hl.record<"Student">>>, !hl.int) -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">>>
      %7 = "hl.implicit_cast"(%6) {kind = 5 : i64} : (!hl.lvalue<!hl.elaborated<!hl.record<"Student">>>) -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">>>
      %8 = "hl.call"(%1, %7) {callee = @_ZN7StudentaSERKS_} : (!hl.lvalue<!hl.elaborated<!hl.record<"Student">>>, !hl.lvalue<!hl.elaborated<!hl.record<"Student">>>) -> !hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">>>>
      "hl.return"() : () -> ()
    }) : () -> ()
  }) {function_type = (!hl.lvalue<!hl.ptr<!hl.elaborated<!hl.record<"Student">>>>, !hl.lvalue<!hl.int>) -> (), linkage = 0 : i32, sym_name = "_Z12sortRegistryP7Studenti"} : () -> ()
}) {dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<!hl.int, 32 : i32>, #dlti.dl_entry<!hl.ptr<!hl.elaborated<!hl.record<"Student">>>, 64 : i32>, #dlti.dl_entry<!hl.ptr<!hl.void>, 64 : i32>, #dlti.dl_entry<!hl.char, 8 : i32>, #dlti.dl_entry<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">>>>, 64 : i32>, #dlti.dl_entry<!hl.void, 0 : i32>, #dlti.dl_entry<!hl.elaborated<!hl.record<"Student">,  const >, 352 : i32>, #dlti.dl_entry<!hl.long< unsigned >, 64 : i32>, #dlti.dl_entry<!hl.array<40, !hl.char>, 320 : i32>, #dlti.dl_entry<!hl.ptr<!hl.void< const >>, 64 : i32>, #dlti.dl_entry<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>, 64 : i32>, #dlti.dl_entry<!hl.elaborated<!hl.record<"Student">>, 352 : i32>, #dlti.dl_entry<!hl.ptr<!hl.record<"Student">>, 64 : i32>, #dlti.dl_entry<!hl.int< const >, 32 : i32>, #dlti.dl_entry<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>, 320 : i32>, #dlti.dl_entry<!hl.void< const >, 0 : i32>, #dlti.dl_entry<!hl.ptr<!hl.adjusted<!hl.array<40, !hl.char>, !hl.array<40, !hl.char>>>, 64 : i32>, #dlti.dl_entry<!hl.record<"Student">, 352 : i32>, #dlti.dl_entry<!hl.rvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">>>>, 64 : i32>>, vast.core.target_triple = "x86_64-apple-macosx13.0.0"} : () -> ()
```

Source to reproduce the issue:
```
struct  Student {
	int registry;
	char name[40];
};

void sortRegistry(struct Student *students, int size)
{
   struct Student aux;
   aux = students[size];
}
```

**Comment by kumarak** [comment]
I see it when lowering the source code with high-level(`hl`) dialects. The `CXXConstructExpr` node is not supported yet but in this case it is lifted as unsupported op. 
```
%9 = "hl.unsupportedOp"() {name = "CXXConstructExpr"} : () -> !hl.elaborated<!hl.record<"Student">>
```

The problem I see is with the lowering of Function type of copy constructor `_ZN7StudentaSERKS_` that has argument type `!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>`. 

Issue is due to type conversion here that is happening due to argument types and return.
```
%10 = "hl.ref"(%arg0) : (!hl.lvalue<!hl.lvalue<!hl.reference<!hl.elaborated<!hl.record<"Student">,  const >>>>) -> !hl.lvalue<!hl.elaborated<!hl.record<"Student">,  const >>
```

**Comment by Jezurko** [comment]
This seems like a similar issue to what I encountered and fixed in #358. Might be worth checking if it's okay now.

**Comment by kumarak** [comment]
@Jezurko, The issue seems to be fixed with the PR you mentioned. The unit test works now. Closing it
